MhLabs.PubSubExtensions/Consumer/Extractors/DynamoDBExtractor.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.DynamoDBv2;
+using Amazon.DynamoDBv2.DataModel;
+using Amazon.DynamoDBv2.DocumentModel;
+using Amazon.Lambda.DynamoDBEvents;
+using MhLabs.PubSubExtensions.Model;
+
+namespace MhLabs.PubSubExtensions.Consumer.Extractors
+{
+    public class DynamoDBExtractor : IMessageExtractor
+    {
+        private readonly IAmazonDynamoDB _dynamoDb;
+
+        public DynamoDBExtractor(IAmazonDynamoDB dynamoDb)
+        {
+            _dynamoDb = dynamoDb;
+        }
+
+        public Type ExtractorForType => typeof(DynamoDBEvent);
+        public async Task<IEnumerable<TMessageType>> ExtractEventBody<TEventType, TMessageType>(TEventType ev) where TMessageType : class, new()
+        {
+            var dynamoEvent = ev as DynamoDBEvent;
+            await Task.CompletedTask;
+
+            using (var context = new DynamoDBContext(_dynamoDb))
+            {
+                return dynamoEvent.Records.Select(record =>
+                {
+                    var newDoc = Document.FromAttributeMap(record.Dynamodb.NewImage);
+                    var oldDoc = Document.FromAttributeMap(record.Dynamodb.OldImage);
+
+                    var update = new MutationModel<TMessageType>
+                    {
+                        OldImage = context.FromDocument<TMessageType>(oldDoc),
+                        NewImage = context.FromDocument<TMessageType>(newDoc)
+                    };
+
+                    return update as TMessageType;
```

Comment:
Does this work?

You pass in a `TMessageType` and return a `MutationModel<TMessageType> as TMessageType`

Surely you'll end up getting an `IEnumerable<TMEssageType>` full of null values?

Say `TMessageType` is `Order`. This code effectively says that `typeof(MutationModel<Order>)` == `typeof(Order)`

If your TMessageType is, as it shoudl be, `MutationModel<Order>`, then you'll try to map the `OldImage` and `NewImage`, which are `Order`s to `MutationModel<Order>`.

To achieve this abstraction you'd need to 'unwrap' `TMessageModel`'s generic argument and use a non-generic version of `_context.FromDocument(...)`, which doesn't exist.

This is why I've so far have resorted to having a local implementation of `DynamoDBExtractor`, because then I know the generic argument type.

I still think we can and should abstract this, but I  think we need to rethink the IMessageExtractor interface.